packages/@react-spectrum/tabs/src/Tabs.tsx

```diff
@@ -34,64 +37,124 @@ export function Tabs<T extends object>(props: SpectrumTabsProps<T>) {
     density,
     ...otherProps
   } = props;
+
   let ref = useRef<HTMLDivElement>();
+  let wrapperRef = useRef<HTMLDivElement>();
   let state = useSingleSelectListState<T>({
     ...props,
     onSelectionChange
   });
 
+  let {direction} = useLocale();
   let {styleProps} = useStyleProps(otherProps);
   let {tabListProps, tabPanelProps} = useTabs(props, state, ref);
+  let [collapse, setCollapse] = useValueEffect(false);
   let [selectedTab, setSelectedTab] = useState<HTMLElement>();
 
   useEffect(() => {
-    let tabs: HTMLElement[] = Array.from(ref.current.querySelectorAll('.' + styles['spectrum-Tabs-item']));
-    setSelectedTab(tabs.find((tab) => tab.dataset.key === state.selectedKey));
-  }, [props.children, state.selectedKey]);
+    // Ensure a tab is always selected (in case no selected key was specified or if selected item was deleted from collection)
+    if (state.selectionManager.isEmpty || !state.collection.getItem(state.selectedKey)) {
+      state.selectionManager.replaceSelection(state.collection.getFirstKey());
+    }
+  }, [state]);
+
+  useEffect(() => {
+    if (ref.current) {
+      let tabs: HTMLElement[] = Array.from(ref.current.querySelectorAll('.' + styles['spectrum-Tabs-item']));
+      let selectedTab = tabs.find((tab) => tab.dataset.key === state.selectedKey);
+
+      if (selectedTab != null) {
+        setSelectedTab(selectedTab);
+      }
+    }
+  }, [props.children, state, ref]);
+
+  let checkShouldCollapse = useCallback(() => {
+    let computeShouldCollapse = () => {
+      if (wrapperRef.current) {
+        let tabsComponent = wrapperRef.current;
+        let tabs = Array.from(ref.current.children);
+        // The last child is the TabLine so we look at the second to last child for the last Tab
+        let lastTab = tabs[tabs.length - 2];
+
+        let end = direction === 'rtl' ? 'left' : 'right';
+        let farEdgeTabList = tabsComponent.getBoundingClientRect()[end];
+        let farEdgeLastTab = lastTab?.getBoundingClientRect()[end];
+        let shouldCollapse = direction === 'rtl' ? farEdgeLastTab < farEdgeTabList : farEdgeTabList < farEdgeLastTab;
+
+        return shouldCollapse;
+      }
+    };
+
+    if (orientation !== 'vertical') {
+      setCollapse(function* () {
+        // Make Tabs render in non-collapsed state
+        yield false;
+
+        // Compute if Tabs should collapse and update
+        yield computeShouldCollapse();
+      });
+    }
+  }, [ref, wrapperRef, direction, orientation, setCollapse]);
+
+  useEffect(() => {
+    checkShouldCollapse();
+  }, [props.children, checkShouldCollapse]);
+
+  useResizeObserver({ref: wrapperRef.current && wrapperRef, onResize: checkShouldCollapse});
```

Comment:
interesting, i didn't have any issue with this one just passing wrapperRef and not checking if current wasn't null
I take it you did?

Reply:
I think this is a hold over from before I tweaked useResizeObserver, removed it and didn't see any issues

Reply:
but if I remember correctly the reproduction was doing the orientation switch from vertical to horizontal